SerializableItem: implement JSON from string

tigercosmos opened the pull request:
Support JSON from string
- add `from_json` into `MM_DECL_SERIALIZABLE`
- implemented `JsonNode`, which is the token of JSON
- implemented JSON parser: `parse_array`, `parse_object`
- added a helper class: `JsonHelper`

tigercosmos committed:
implement json from string

tigercosmos commented on cpp/modmesh/base.hpp:
```diff
@@ -72,6 +72,30 @@ struct is_specialization_of<Template, Template<Args...>> : std::true_type
 template <template <typename...> class Template, typename T>
 inline constexpr bool is_specialization_of_v = is_specialization_of<Template, T>::value;
 
+inline bool is_alpha(char c)
```
some common helper functions

yungyuc replied:
`isalpha`: https://en.cppreference.com/w/cpp/string/byte/isalpha

tigercosmos replied:
thanks for telling me

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -27,13 +27,17 @@
 namespace detail
 {
 
+inline void throw_serialization_error(Formatter & formatter, int line, int column)
```
print readable error message for parsing json

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -76,6 +80,460 @@ std::string escape_string(std::string_view str_view)
+    return (start == std::string::npos) ? "" : str.substr(start, end - start + 1);
+}
+
+#define MM_DECL_CACULATE_LINE_COLUMN(CH) \
```
for the error message

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -76,6 +80,460 @@ std::string escape_string(std::string_view str_view)
+        column += 1;                     \
+    }
+
+JsonArray JsonNode::parse_array(const std::string & json)
```
a state machine parsing JSON array, same for the `parse_object`

yungyuc replied:
Sweet. I like it. It's the kind of code I cannot quickly write.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -76,6 +80,460 @@ std::string escape_string(std::string_view str_view)
+                    throw_serialization_error(Formatter() << "Invalid JSON format: missing closing bracket.", line, column);
+                }
+
+                json_array.emplace_back(std::make_unique<JsonNode>(JsonType::Object, value_expression));
```
parse the child, which is a JSON object here.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -76,6 +80,460 @@ std::string escape_string(std::string_view str_view)
+                    throw_serialization_error(Formatter() << "Invalid JSON format: missing closing bracket.", line, column);
+                }
+
+                json_array.emplace_back(std::make_unique<JsonNode>(JsonType::Array, value_expression));
```
the child is array.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -76,6 +80,460 @@ std::string escape_string(std::string_view str_view)
+                    }
+                }
+
+                json_array.emplace_back(std::make_unique<JsonNode>(type, value_expression));
```
the basic types go to here.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -76,6 +80,433 @@ std::string escape_string(std::string_view str_view)
+            else if (c == '[')
+            {
+                // array in array is valid in JSON, but we do not support it
+                throw_serialization_error(Formatter() << "Invalid JSON format: unexpected array in array.", line, column);
```
not support array in array for now.

yungyuc replied:
It'll be good to append "yet" in the code comment.  Nested arrays will be needed.

I agree we do not need it badly at the moment.

tigercosmos replied:
fixed.

yungyuc replied:
This is good. Thanks.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -76,6 +80,433 @@ std::string escape_string(std::string_view str_view)
+    return json_array;
+}
+
+JsonMap JsonNode::parse_object(const std::string & json)
```
state machine for JSON object.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -76,6 +80,433 @@ std::string escape_string(std::string_view str_view)
+                    throw_serialization_error(Formatter() << "Invalid JSON format: missing closing bracket.", line, column);
+                }
+
+                json_map.emplace(key, std::make_unique<JsonNode>(JsonType::Object, value_expression));
```
nested object.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -76,6 +80,433 @@ std::string escape_string(std::string_view str_view)
+                    throw_serialization_error(Formatter() << "Invalid JSON format: missing closing bracket.", line, column);
+                }
+
+                json_map.emplace(key, std::make_unique<JsonNode>(JsonType::Array, value_expression));
```
the child is array.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -55,65 +56,265 @@ namespace detail
+}; /* end enum class JsonState */
+
+/// Type of JSON token.
+enum class JsonType
```
support common types for now.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -55,65 +56,265 @@ namespace detail
+{
+public:
+    template <typename T>
+    static std::string to_json_string(const T & value)
```
move this function into the helper class.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -55,65 +56,265 @@ namespace detail
-    else if constexpr (is_specialization_of<std::vector, T>::value)
+
+    template <typename T>
+    static void from_json_string(const detail::JsonNodePtr & node, T & value)
```
save the value into the class.

yungyuc replied:
I don't like the output argument.  But it looks like necessary evil here.  Is it so?  If yes, is it possible to make it private?

yungyuc replied:
And this function `from_json_string` is certainly too long for declaration.  Please make it outside the class declaration.

tigercosmos replied:
> And this function from_json_string is certainly too long for declaration. Please make it outside the class declaration.

I put it in the class because L228 (`value.from_json(std::get<JsonMap>(node->value))`) uses the private method of the class that inherits `SerializableItem`, and that's why I declare `friend class detail::JsonHelper;`

tigercosmos replied:
> I don't like the output argument. But it looks like necessary evil here. Is it so? If yes, is it possible to make it private?

based on the current design, it is necessary.
What do you mean "private"? this is a helper class, and the function is static.
`SerializableItem::from_json` calls this function, and feeds the class member throw the output argument.

tigercosmos replied:
In addition, since it's a helper class (or basically a helper function), I think the length should be fine? we have much longer helper class in `array_common.hpp`

yungyuc replied:
> > And this function from_json_string is certainly too long for declaration. Please make it outside the class declaration.
> 
> I put it in the class because L228 (when the parser doing recursion) uses the private method of `SerializableItem`, and that's why I declare `friend class detail::JsonHelper;`

The access is OK.  Long functions are not easy to read when putting in class declaration.  They should be taken outside the declaration while still made inline.

yungyuc replied:
> > I don't like the output argument. But it looks like necessary evil here. Is it so? If yes, is it possible to make it private?
> 
> based on the current design, it is necessary. What do you mean "private"? this is a helper class, and the function is static. `SerializableItem::from_json` calls this function, and feeds the class member throw the output argument.

It will be easier to maintain a function of output argument when it is private.  It does not matter the function to be static or not.  But if you use the class as a namespace for collecting static member functions and call them from other class, making them private is not realistic.

Please leave a code comment to warn maintainers of the output argument.  We should get rid of it in the future.

tigercosmos replied:
@yungyuc  based on the current design:
in `from_json_string`:
```cpp
    else if constexpr (std::is_base_of_v<SerializableItem, T>)
    {
        if (node->type != detail::JsonType::Object)
        {
            throw std::runtime_error("Invalid JSON format: invalid object type.");
        }
        // NOLINTNEXTLINE(misc-no-recursion)
        value.from_json(std::get<JsonMap>(node->value)); /* recursive here */
    }
 ```
 
Note that we call `value`'s mutable method `from_json`. Therefore, we must pass the **mutable** reference of `value` as the argument of `from_json_string`.
 
 I don't think it can be an easy change to get rid of the output argument unless we change the whole design.
 
 Also,  I didn't fully understand why you wanted to remove the output argument. This function is live in the `detail` namespace, and the implementation is hidden under the macro, I don't think the user has chance to use this function.

yungyuc replied:
It's a good point that the static member function is in `detail`.  I missed it.  In this case I agree the output argument is not so severe.  Could you please add a simple code comment indicating the output arguments are bad for maintenance and should be fixed in the future?  (A concrete plan does not need to be mentioned.)

It's quite possible for a maintainer to move a useful helper class outside `detail`, in the future.  Then it will become an issue.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -55,65 +56,265 @@ namespace detail
+                                                                                                    \
+    void from_json(const std::string & json) override                                               \
+    {                                                                                               \
+        auto jsonNode = std::make_unique<detail::JsonNode>(detail::JsonType::Object, json);         \
```
parse the JSON first.

tigercosmos commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -55,65 +56,265 @@ namespace detail
+                auto it = obj.find(name);                                                           \
+                if (it != obj.end())                                                                \
+                {                                                                                   \
+                    detail::JsonHelper::from_json_string(it->second, value);                        \
```
then we can save the values from the parsed nodes into class's values.

yungyuc commented on cpp/modmesh/base.hpp:
```diff
@@ -72,6 +72,30 @@ struct is_specialization_of<Template, Template<Args...>> : std::true_type
+    return c == ' ' || c == '\t' || c == '\n' || c == '\r';
+}
+
+template <typename T>
```
Use STL `is_integral`: https://en.cppreference.com/w/cpp/types/is_integral

yungyuc commented on cpp/modmesh/base.hpp:
```diff
@@ -72,6 +72,30 @@ struct is_specialization_of<Template, Template<Args...>> : std::true_type
+                              std::is_same_v<T, uint32_t> || std::is_same_v<T, uint64_t>;
+
+template <typename T>
+constexpr bool is_float_v = std::is_same_v<T, float> || std::is_same_v<T, double>;
```
`is_floating_point`: https://en.cppreference.com/w/cpp/types/is_floating_point

yungyuc commented on cpp/modmesh/base.hpp:
```diff
@@ -72,6 +72,30 @@ struct is_specialization_of<Template, Template<Args...>> : std::true_type
+    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
+}
+
+inline bool is_digit(char c)
```
`isdigit`: https://en.cppreference.com/w/cpp/string/byte/isdigit

yungyuc commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -27,13 +27,17 @@
  */
 
 #include <modmesh/serialization/SerializableItem.hpp>
```
Thanks for adding the blank line.

tigercosmos replied:
sorry, fixed.

tigercosmos replied:
I hope clang-format can help for this.

yungyuc replied:
If it does not, we can consider to craft a linter for ourselves.  In the future, of course.

yungyuc commented on cpp/modmesh/serialization/SerializableItem.cpp:
```diff
@@ -76,6 +80,433 @@ std::string escape_string(std::string_view str_view)
+            }
+
+            state = JsonState::ObjectKey;
+            break; /* end case JsonState::Start */
```
I like the case ending marks.

yungyuc commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -55,65 +56,265 @@ namespace detail
+}; /* end enum class JsonType */
+
+struct JsonNode; // Forward declaration
+typedef std::unique_ptr<JsonNode> JsonNodePtr;
```
Do not add type aliases to smart pointers (`JsonNodePtr`).  Use the modern style `using` (for style consistency):

```cpp
using std::unordered_map<std::string, std::unique_ptr<JsonNode>> JsonMap;
using JsonArray = std::vector<std::unique_ptr<JsonNode>>;
```

yungyuc replied:
Aliases to smart pointer types always confuse people.  We will need to look up the definition.  It's better to always spell out the smart pointer type.

tigercosmos replied:
fixed

yungyuc commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -55,65 +56,265 @@ namespace detail
+    void parse(const std::string & expression)
     {
-        return "\"" + escape_string(value) + "\"";
+        if (type == JsonType::Object)
```
The function body of `parse` is a little bit too long for declaration, but it's fine for now.

yungyuc commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -55,65 +56,265 @@ namespace detail
+            oss << "]";
+            return oss.str();
+        }
+        else
```
I am not familiar with `constexpr if` enough.  Do you need the keyword here?

tigercosmos replied:
`constexpr if` is doing the condition in compilation time, which will not have a runtime branch.

yungyuc replied:
My question is against `else`.  Does it need `constexpr`?

tigercosmos replied:
> My question is against else. Does it need constexpr?

No need.

yungyuc commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -55,65 +56,265 @@ namespace detail
-    else
+
+    template <typename T>
+    static void from_json_string(const JsonNodePtr & node, std::vector<T> & vec)
```
This overload can be outside the declaration too.

tigercosmos replied:
this function is okay to put outside, since it doesn't use the private function. However, as `static void from_json_string(const detail::JsonNodePtr & node, T & value)` must be put in the helper class, I think we should put this function in the helper class for the consistency?

yungyuc replied:
By putting outside the (class) declaration, I meant to move the function body, like:

```cpp
class C
{
    void F();
};
void C::F()
{
    // do something
}
```

It has nothing to do with function access.

tigercosmos replied:
I see, I realize it's possible to do it even for templates.

yungyuc commented on gtests/test_nopython_serializable.cpp:
```diff
@@ -163,4 +167,111 @@ TEST(Json, serialization_with_object)
                          "\"pets\":[{\"name\":\"Fluffy\",\"age\":3,\"is_dog\":true,\"is_cat\":false},{\"name\":\"Whiskers\",\"age\":8,\"is_dog\":false,\"is_cat\":true}]}";
 }
 
+TEST(Json, deserialize_private_member_partial_exposure)
```
We will need Python tests for these things.  It's not productive to test for strings in C++.

It can use a new PR or even a new issue.

tigercosmos replied:
sure, should be in the future PRs.

yungyuc reviewed (changes requested):
It's an exciting PR.  Things to fix:

- [x] Use STL `isalpha`, `isdigit`, `is_integral`, and `is_floating_point`.
- [x] Do not alias smart pointer types.
- [x] Use the modern `using` syntax.
- [x] Clarify how to improve the output argument of `from_json_string()`.
- [x] `from_json_string()` is too long for class declaration.

tigercosmos commented:
@yungyuc 
> Use STL isalpha, isdigit, is_integral, and is_floating_point.

all fixed.

>  Do not alias smart pointer types.

fixed.

> Use the modern using syntax.

fixed.

> Clarify how to improve the output argument of from_json_string().

output argument is needed by design.

> from_json_string() is too long for class declaration.

splitted.

All issues are addressed, please also check the comments.

yungyuc commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -89,31 +174,160 @@ std::string to_json_string(const T & value)
 
-}; /* end namespace detail */
+template <typename T>
+void JsonHelper::from_json_string(const std::unique_ptr<JsonNode> & node, T & value)
```
Need code comment for the plan to mitigate the output arguments.

yungyuc replied:
(I hit review changes button too quickly.)  We can discuss here or discord for how to author the plan.  It may be a simple one.

yungyuc commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -89,31 +174,160 @@ std::string to_json_string(const T & value)
+}
+
+template <typename T>
+void JsonHelper::from_json_string(const std::unique_ptr<JsonNode> & node, std::vector<T> & vec)
```
Ditto.  Output argument.

yungyuc reviewed (changes requested):
- [x] Leave code comments for a plan to mitigate the output argument.

tigercosmos commented:
@yungyuc  please see my comment here: https://github.com/solvcon/modmesh/pull/377#discussion_r1653774032

tigercosmos committed:
add TODO comments

tigercosmos commented:
@yungyuc  comments for the output argument is added.

yungyuc commented on cpp/modmesh/serialization/SerializableItem.hpp:
```diff
@@ -55,11 +56,94 @@ namespace detail
+    static std::string to_json_string(const T & value);
+
+    template <typename T>
+    static void from_json_string(const std::unique_ptr<JsonNode> & node, T & value); // TODO: have a design that can remove the output argument to increase the maintainability
```
Thanks, but please beautify the comments.  Appending after the signature makes the lines too line.  Please add the comments before the signature in distinct lines.

yungyuc reviewed (changes requested):
Thanks, @tigercosmos .  We need some cosmetics.

- [x] Cosmetic changes for the comments.

tigercosmos committed:
TODO

yungyuc commented:
I saw the change and will merge once CI completes.
